File: app/src/main/java/com/example/david/lists/widget/view/WidgetRemoteView.java
package com.example.david.lists.widget.view;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.widget.RemoteViews;

import com.example.david.lists.R;
import com.example.david.lists.view.itemlist.ItemActivity;
import com.example.david.lists.widget.configview.WidgetConfigView;

import static com.example.david.lists.util.UtilWidgetKeys.getSharedPrefKeyId;
import static com.example.david.lists.util.UtilWidgetKeys.getSharedPrefKeyTitle;
import static com.example.david.lists.util.UtilWidgetKeys.getSharedPrefName;

public final class WidgetRemoteView {

    private final RemoteViews remoteViews;
    private final int appWidgetId;

    private final Context context;

    public WidgetRemoteView(Context context, int appWidgetId) {
        this.remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget);
        this.appWidgetId = appWidgetId;
        this.context = context;
    }


    public RemoteViews updateWidget() {
        initRemoveView();
        return remoteViews;
    }

    private void initRemoveView() {
        setTitle();
        setTitlePendingIntent();
        setConfigActivityPendingIntent();
        initRemoteAdapter();
    }


    private void setTitlePendingIntent() {
        Intent intent = new Intent(context, ItemActivity.class);
        setIntentExtras(intent);
        remoteViews.setOnClickPendingIntent(
                R.id.widget_tv_title,
                getPendingIntent(intent)
        );
    }

    private void setIntentExtras(Intent intent) {
        intent.putExtra(getString(R.string.intent_extra_user_list_id), getListId());
        intent.putExtra(getString(R.string.intent_extra_user_list_title), getListTitle());
    }

    private void setTitle() {
        remoteViews.setTextViewText(R.id.widget_tv_title, getListTitle());
    }

    private String getListTitle() {
        return getSharedPrefs().getString(getSharedPrefKeyTitle(context, appWidgetId), null);
    }


    private void setConfigActivityPendingIntent() {
        Intent intent = new Intent(context, WidgetConfigView.class);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        remoteViews.setOnClickPendingIntent(
                R.id.widget_iv_settings,
                getPendingIntent(intent)
        );
    }


    private void initRemoteAdapter() {
        Intent adapterIntent = new Intent(context, MyRemoteViewsService.class);
        setAdapterIntentExtras(adapterIntent);
        remoteViews.setRemoteAdapter(R.id.widget_list_view, adapterIntent);
        remoteViews.setEmptyView(R.id.widget_list_view, R.id.widget_tv_error);
    }

    private void setAdapterIntentExtras(Intent adapterIntent) {
        adapterIntent.putExtra(
                context.getString(R.string.intent_extra_user_list_id), getListId()
        );
        adapterIntent.putExtra(
                AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId
        );
        // So the system can distinguish between the different widget instances
        adapterIntent.setData(
                Uri.parse(adapterIntent.toUri(Intent.URI_INTENT_SCHEME))
        );
    }

    private String getListId() {
        return getSharedPrefs().getString(getSharedPrefKeyId(context, appWidgetId), "");
    }


    private PendingIntent getPendingIntent(Intent intent) {
        return PendingIntent.getActivity(
                context, appWidgetId, intent, PendingIntent.FLAG_UPDATE_CURRENT
        );
    }

    private SharedPreferences getSharedPrefs() {
        return context.getSharedPreferences(
                getSharedPrefName(context),
                Context.MODE_PRIVATE
        );
    }


    private String getString(int stringRes) {
        return context.getString(stringRes);
    }
}

File: app/src/main/java/com/precopia/david/lists/widget/view/WidgetRemoteView.kt
package com.precopia.david.lists.widget.view

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import com.precopia.david.lists.R
import com.precopia.david.lists.widget.view.IWidgetRemoteViewContract.Companion.INTENT_ADAPTER
import com.precopia.david.lists.widget.view.IWidgetRemoteViewContract.Companion.INTENT_CONFIG
import com.precopia.david.lists.widget.view.IWidgetRemoteViewContract.Companion.INTENT_TITLE
import com.precopia.david.lists.widget.view.IWidgetRemoteViewContract.Companion.LIST_TITLE
import com.precopia.david.lists.widget.view.buildlogic.DaggerWidgetRemoteViewComponent
import javax.inject.Inject
import javax.inject.Named

class WidgetRemoteView(private val context: Context,
                       private val widgetId: Int) {

    init {
        DaggerWidgetRemoteViewComponent.builder()
                .context(context)
                .widgetId(widgetId)
                .build()
                .inject(this)
    }

    @Inject
    lateinit var remoteViews: RemoteViews

    @Inject
    @Named(LIST_TITLE)
    lateinit var listTitle: String

    @Inject
    @Named(INTENT_TITLE)
    lateinit var titleIntent: PendingIntent

    @Inject
    @Named(INTENT_CONFIG)
    lateinit var configActivityIntent: Intent

    @Inject
    @Named(INTENT_ADAPTER)
    lateinit var adapterIntent: Intent


    fun updateWidget(): RemoteViews {
        initRemoveView()
        return remoteViews
    }

    private fun initRemoveView() {
        setTitle()
        setTitlePendingIntent()
        setConfigActivityPendingIntent()
        initRemoteAdapter()
    }

    private fun setTitle() {
        remoteViews.setTextViewText(R.id.widget_tv_title, listTitle)
    }

    private fun setTitlePendingIntent() {
        remoteViews.setOnClickPendingIntent(
                R.id.widget_tv_title,
                titleIntent
        )
    }

    private fun setConfigActivityPendingIntent() {
        remoteViews.setOnClickPendingIntent(
                R.id.widget_iv_settings,
                getPendingIntent(configActivityIntent)
        )
    }

    private fun getPendingIntent(intent: Intent) = PendingIntent.getActivity(
            context, widgetId, intent, PendingIntent.FLAG_UPDATE_CURRENT
    )

    private fun initRemoteAdapter() {
        remoteViews.setRemoteAdapter(R.id.widget_list_view, adapterIntent)
        remoteViews.setEmptyView(R.id.widget_list_view, R.id.widget_tv_error)
    }
}
